### PR TITLE
feat: implement typed events

### PR DESCRIPTION
**Pull request by jrandolf-google**
This PR implements typings for emitted events in all event emitters.

### CODE REVIEW

File: packages/puppeteer-core/src/api/Frame.ts [hunk 1]
```diff
@@ -127,6 +127,44 @@ export interface FrameAddStyleTagOptions {
   content?: string;
 }
 
+/**
+ * @public
+ */
+export type FrameEvents = {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/api/Frame.ts]: Why make the type public but every entry is internal?

File: packages/puppeteer-core/src/api/Page.ts [hunk 2]
```diff
@@ -33,20 +33,25 @@ import {
 import type {HTTPRequest} from '../api/HTTPRequest.js';
 import type {HTTPResponse} from '../api/HTTPResponse.js';
 import type {Accessibility} from '../common/Accessibility.js';
-import type {CDPSession} from '../common/Connection.js';
+import type {BidiNetworkManager as BidiNetworkManager} from '../common/bidi/NetworkManager.js';
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/api/Page.ts]: Unnecessary name change.

File: packages/puppeteer-core/src/api/Page.ts [hunk 3]
```diff
@@ -513,54 +535,48 @@ export abstract class Page
...
-   * Listen to page events.
-   *
-   * :::note
-   *
-   * This method exists to define event typings and handle proper wireup of
-   * cooperative request interception. Actual event listening and dispatching is
-   * delegated to {@link EventEmitter}.
-   *
-   * :::
+   * @internal
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/api/Page.ts:550]: I think comment is still relevant.

File: packages/puppeteer-core/src/common/CDPSession.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  createProtocolErrorMessage,
+} from './Connection.js';
+import {TargetCloseError} from './Errors.js';
+import {CDPTarget} from './Target.js';
+
+/**
+ * @internal
+ */
+
+export class CDPCDPSession extends CDPSession {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/CDPSession.ts:18]: Use the common prefix for specific `Cdp`/`Bidi` implementations
```suggestion
export class CdpCDPSession extends CDPSession {
```

File: packages/puppeteer-core/src/common/ExecutionContext.ts [hunk 5]
```diff
@@ -47,25 +47,6 @@ const getSourceUrlComment = (url: string) => {
 };
 
 /**
- * Represents a context for JavaScript execution.
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/ExecutionContext.ts:50]: Keep comments.

File: packages/puppeteer-core/src/common/FrameManager.ts [hunk 6]
```diff
@@ -190,7 +202,7 @@ export class FrameManager extends EventEmitter {
     });
     session.on('Page.frameNavigated', event => {
       this.#frameNavigatedReceived.add(event.frame.id);
-      void this.#onFrameNavigated(event.frame, event.type);
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/FrameManager.ts]: Why this change?

File: packages/puppeteer-core/src/common/FrameManager.ts [hunk 7]
```diff
@@ -346,7 +358,7 @@ export class FrameManager extends EventEmitter {
       );
     }
     if (!this.#frameNavigatedReceived.has(frameTree.frame.id)) {
-      void this.#onFrameNavigated(frameTree.frame, 'Navigation');
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/FrameManager.ts]: Why this change?

File: packages/puppeteer-core/src/common/FrameManager.ts [hunk 8]
```diff
@@ -412,8 +421,8 @@ export class FrameManager extends EventEmitter {
 
     frame = await this._frameTree.waitForFrame(frameId);
     frame._navigated(framePayload);
-    this.emit(FrameManagerEmittedEvents.FrameNavigated, frame);
-    frame.emit(FrameEmittedEvents.FrameNavigated, navigationType);
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/FrameManager.ts:416]: Why this change?

File: packages/puppeteer-core/src/common/FrameManager.ts [hunk 9]
```diff
@@ -474,8 +483,9 @@ export class FrameManager extends EventEmitter {
         this.#removeFramesRecursively(frame);
       }
     } else if (reason === 'swap') {
-      this.emit(FrameManagerEmittedEvents.FrameSwapped, frame);
-      frame?.emit(FrameEmittedEvents.FrameSwapped);
+      assert(frame);
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/FrameManager.ts]: This can produce unhandled rejection. Wrap like the other?

File: packages/puppeteer-core/src/common/HTTPRequest.ts [hunk 10]
```diff
@@ -30,20 +31,19 @@ import {
...
 
-import {CDPSession} from './Connection.js';
 import {ProtocolError} from './Errors.js';
 import {debugError, isString} from './util.js';
 
 /**
  * @internal
  */
-export class HTTPRequest extends BaseHTTPRequest {
+export class CDPHTTPRequest extends HTTPRequest {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/HTTPRequest.ts:40]:
```suggestion
export class CdpHTTPRequest extends HTTPRequest {
```

File: packages/puppeteer-core/src/common/HTTPResponse.ts [hunk 11]
```diff
@@ -15,24 +15,21 @@
...
 import {ProtocolError} from './Errors.js';
-import {HTTPRequest} from './HTTPRequest.js';
+import {CDPHTTPRequest} from './HTTPRequest.js';
 import {SecurityDetails} from './SecurityDetails.js';
 
 /**
  * @internal
  */
-export class HTTPResponse extends BaseHTTPResponse {
+export class CDPHTTPResponse extends HTTPResponse {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/HTTPResponse.ts:30]:
```suggestion
export class CdpHTTPResponse extends HTTPResponse {
```

File: packages/puppeteer-core/src/common/NetworkManager.test.ts [hunk 12]
```diff
@@ -692,14 +692,11 @@ describe('NetworkManager', () => {
 
     const finishedRequests: HTTPRequest[] = [];
     const pendingRequests: HTTPRequest[] = [];
-    manager.on(
-      NetworkManagerEmittedEvents.RequestFinished,
-      (request: HTTPRequest) => {
-        finishedRequests.push(request);
-      }
-    );
+    manager.on(NetworkManagerEvent.RequestFinished, (request: HTTPRequest) => {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/NetworkManager.test.ts:695]: Is type here for property still needed?

File: packages/puppeteer-core/src/common/Page.ts [hunk 13]
```diff
@@ -146,58 +142,59 @@ export class CDPPage extends Page {
   #serviceWorkerBypassed = false;
   #userDragInterceptionEnabled = false;
 
-  #frameManagerHandlers = new Map<symbol, Handler<any>>([
+  #frameManagerHandlers = Object.freeze([
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/Page.ts:145]: Maybe we should use `readonly` as they are internal methods.

File: packages/puppeteer-core/src/common/Page.ts [hunk 14]
```diff
@@ -146,58 +142,59 @@ export class CDPPage extends Page {
...
+        this.#sessionCloseDeferred.resolve(
           new TargetCloseError('Target closed')
         );
       },
     ],
     [
       'Page.domContentEventFired',
-      this.emit.bind(this, PageEmittedEvents.DOMContentLoaded),
+      () => {
+        return this.emit(PageEvent.DOMContentLoaded, undefined);
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/Page.ts:211]: Why this change, can't you pass `undefined` as 3th parameter?

File: packages/puppeteer-core/src/common/Page.ts [hunk 15]
```diff
@@ -254,57 +251,50 @@ export class CDPPage extends Page {
...
-          return;
-        }
-        this.#frameManager
-          .registerSpeculativeSession(session)
-          .catch(debugError);
-        this.#emulationManager
-          .registerSpeculativeSession(session)
-          .catch(debugError);
+    this.#tabSession?.on(CDPSessionEvent.Ready, session => {
+      assert(session instanceof CDPCDPSession);
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/Page.ts:277]: Not sure of this check.

File: packages/puppeteer-core/src/common/WebWorker.ts [hunk 16]
```diff
@@ -69,7 +68,7 @@ export type ExceptionThrownCallback = (
  *
  * @public
  */
-export class WebWorker extends EventEmitter {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/WebWorker.ts:72]: This is a breaking change.

File: packages/puppeteer-core/src/common/bidi/BidiOverCDP.ts [hunk 17]
```diff
@@ -53,7 +50,7 @@ export async function connectBidiOverCDP(
     // Forwards a BiDi event sent by BidiServer to Puppeteer.
     pptrTransport.onmessage(JSON.stringify(message));
   });
-  const pptrBiDiConnection = new BidiPPtrConnection(cdp.url(), pptrTransport);
+  const pptrBiDiConnection = new BidiConnection(cdp.url(), pptrTransport);
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/bidi/BidiOverCDP.ts:53]: Not sure of the renaming to `BidiConnection` will have to look what the class does.

File: packages/puppeteer-core/src/common/bidi/Connection.ts [hunk 18]
```diff
@@ -218,7 +225,7 @@ export class Connection extends EventEmitter {
     } else if (isCDPEvent(event)) {
       cdpSessions
         .get(event.params.session)
-        ?.emit(event.params.event, event.params.params);
+        ?.emit(event.params.event, event.params.params[0]);
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/bidi/Connection.ts]: Why this change? I think that's not relevant for BiDi+

File: packages/puppeteer-core/src/api/Frame.ts (hunk 1 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/api/Frame.ts]: The type cannot resolve in the public types if `FrameEvents` is internal. This will lead to an unknown signature for `Frame`.

File: packages/puppeteer-core/src/common/CDPSession.ts (hunk 4 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/CDPSession.ts:18]: We can do this is another PR. We should consistently use `CDP` since that's how it is everywhere in puppeteer.

File: packages/puppeteer-core/src/common/bidi/Connection.ts (hunk 18 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/bidi/Connection.ts]: Because that's the correct typing.

File: packages/puppeteer-core/src/common/Page.ts (hunk 15 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/Page.ts:277]: What are you not sure about?

File: packages/puppeteer-core/src/common/Page.ts (hunk 14 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/Page.ts:211]: TypeScript has trouble without the above.

File: packages/puppeteer-core/src/common/ExecutionContext.ts (hunk 5 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/ExecutionContext.ts:50]: This is left-over from when this was public. The comment is not relevant anymore (and we plan to remove this anyway).

**Comment by OrKoN** [comment]
Would it be considered a breaking change? I assume if someone used string literals for event names before it might not compile after this change.

**Comment by OrKoN** [comment]
Could you add a description about what has specifically changed with a basic before and after example? (I assume `feat:` is marked correctly and indicates a user facing feature and not a refactoring).

**Comment by jrandolf** [comment]
> Would it be considered a breaking change? I assume if someone used string literals for event names before it might not compile after this change.

There are no breaking changes, see tests for example.

> Could you add a description about what has specifically changed with a basic before and after example? (I assume feat: is marked correctly and indicates a user facing feature and not a refactoring).

Done.

**Comment by OrKoN** [comment]
> > Would it be considered a breaking change? I assume if someone used string literals for event names before it might not compile after this change.
> 
> There are no breaking changes, see tests for example.
> 
> > Could you add a description about what has specifically changed with a basic before and after example? (I assume feat: is marked correctly and indicates a user facing feature and not a refactoring).
> 
> Done.

so what is the user facing change then? should it be prefixed as `refactor:`?

**Comment by jrandolf** [comment]
> so what is the user facing change then? should it be prefixed as refactor:?

Users now get the correct typing when they try to add a listener which required manual typing previously. We could label it as a `refactor` if you feel this isn't sufficient to label as a `feat`.

**Comment by OrKoN** [comment]
> > so what is the user facing change then? should it be prefixed as refactor:?
> 
> Users now get the correct typing when they try to add a listener which required manual typing previously. We could label it as a `refactor` if you feel this isn't sufficient to label as a `feat`.

I am not sure, maybe feat is good. Btw, GitHub UI really struggles to render the diffs in readable manner: I guess it's lazy-loaded so you cannot actually jump to a specific file without scrolling for a while first.

File: packages/puppeteer-core/src/common/FrameManager.ts (hunk 9 above)
Comment (Lightning00Blade) [reply packages/puppeteer-core/src/common/FrameManager.ts]: Do an early return instead.

File: packages/puppeteer-core/src/common/FrameManager.ts (hunk 6 above)
Comment (Lightning00Blade) [reply packages/puppeteer-core/src/common/FrameManager.ts]: Not addressed, this may brake BFCache

File: packages/puppeteer-core/src/common/bidi/NetworkManager.ts [hunk 19]
```diff
@@ -16,40 +16,71 @@
...
+  [NetworkManagerEvent.RequestFinished]: BidiHTTPRequest;
+}
+
+/**
+ * @internal
+ */
+export class BidiNetworkManager extends EventEmitter<NetworkManagerEvents> {
+  #connection: BidiConnection;
   #page: BidiPage;
-  #subscribedEvents = new Map<string, Handler<any>>([
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/bidi/NetworkManager.ts:34]: It feels quicker to find what events when searching, when we have them in a map and then loop though them. 
We made the change in other places to use the same pattern. 
Leaving to @OrKoN to give a vote on this.

File: packages/puppeteer-core/src/common/bidi/Connection.ts [hunk 20]
```diff
@@ -202,7 +212,8 @@ export class Connection extends EventEmitter {
         debugError(object);
       } else {
         this.#maybeEmitOnContext(object);
-        this.emit(object.method, object.params);
+        // SAFETY: We know the method and parameter still match here.
+        this.emit(object.method, object.params as Bidi.BrowsingContext.Info);
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/bidi/Connection.ts]: Why exactly `Bidi.BrowsingContext.Info` shouldn't it be something more generic? Like `Bidi.ChromiumBidi.Event`?

File: test/src/bfcache.spec.ts [hunk 21]
```diff
@@ -25,7 +25,7 @@ describe('BFCache', function () {
     });
 
     try {
-      page.setDefaultTimeout(2000);
+      page.setDefaultTimeout(5000);
```
Comment (Lightning00Blade) [inline test/src/bfcache.spec.ts]: Why this change?

File: packages/puppeteer-core/src/common/bidi/Context.ts
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/common/bidi/Context.ts:1]: Why is this empty, delete?

File: packages/puppeteer-core/src/common/bidi/NetworkManager.ts (hunk 19 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/common/bidi/NetworkManager.ts:34]: since DisposableStack keep the track of the list of event subscriptions, I think it's okay.

File: test/src/bfcache.spec.ts (hunk 21 above)
Comment (jrandolf) [reply test/src/bfcache.spec.ts]: Tests were timing out here.

File: packages/puppeteer-core/src/common/bidi/Context.ts
Comment (jrandolf) [reply packages/puppeteer-core/src/common/bidi/Context.ts:1]: Probably an accidentally committed file.

File: packages/puppeteer-core/src/common/bidi/Connection.ts (hunk 20 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/common/bidi/Connection.ts]: TypeScript has trouble here because `object.method` is a union of all the method types, so it doesn't know how to interpret `object.params`

File: test/src/bfcache.spec.ts (hunk 21 above)
Comment (OrKoN) [reply test/src/bfcache.spec.ts]: 2 sec should be more than enough for anything in this test unless smth else is going wrong or the host is not healthy

Comment (OrKoN) [reply test/src/bfcache.spec.ts]: it seems it timed out with 5 sec too

File: packages/puppeteer-core/src/common/NetworkManager.test.ts (hunk 12 above)
Comment (Lightning00Blade) [reply packages/puppeteer-core/src/common/NetworkManager.test.ts:695]: Please explain why we still need this.

File: packages/puppeteer-core/src/common/bidi/Connection.ts (hunk 20 above)
Comment (Lightning00Blade) [reply packages/puppeteer-core/src/common/bidi/Connection.ts]: Nit: I see, let's type it as `any` as this is not the correct type, comment explains why we do that.

**Review by Lightning00Blade — APPROVED** [review]
LGTM + please answer the two question like comments and look at nits, and wait for all test to pass.

File: packages/puppeteer-core/src/common/NetworkManager.test.ts (hunk 12 above)
Comment (Lightning00Blade) [reply packages/puppeteer-core/src/common/NetworkManager.test.ts:695]: It looks like we don't need this anymore can we put TODO/Issue to remove them and rely on our Types.
